File: CPP/mergesort/mergesort_250000.cpp
#include <iostream>
#include <vector>
using namespace std;

void merge(vector<int>& arr, int left, int mid, int right) {
    int len1 = mid - left + 1;
    int len2 = right - mid;

    vector<int> leftArr(len1);
    vector<int> rightArr(len2);

    for (int i = 0; i < len1; i++)
        leftArr[i] = arr[left + i];
    for (int j = 0; j < len2; j++)
        rightArr[j] = arr[mid + 1 + j];

    int i = 0, j = 0, k = left;

    while (i < len1 && j < len2) {
        if (leftArr[i] <= rightArr[j]) {
            arr[k] = leftArr[i];
            i++;
        } else {
            arr[k] = rightArr[j];
            j++;
        }
        k++;
    }

    while (i < len1) {
        arr[k] = leftArr[i];
        i++;
        k++;
    }

    while (j < len2) {
        arr[k] = rightArr[j];
        j++;
        k++;
    }
}

void mergeSort(vector<int>& arr, int left, int right) {
    if (left < right) {
        int mid = (left + right) / 2;
        mergeSort(arr, left, mid);
        mergeSort(arr, mid + 1, right);
        merge(arr, left, mid, right);
    }
}

int main() {
    bool debug = false;

    
    int n = arr.size();
    mergeSort(arr, 0, n-1);
    if(debug) {
        for (int i = 0; i < n; i++){
            cout << arr[i] << " ";
        }
    } 

    return 0;
}

File: CSHARP/mergesort/src/mergesort_250000.cs
using System;

class MergeSort{

    public static bool debug = false;

    public static void mergeSort(int[] array)
    {
        if (array.Length > 1)
        {
            int mid = array.Length / 2;

            int[] leftArray = new int[mid];
            int[] rightArray = new int[array.Length - mid];

            for (int i = 0; i < mid; i++)
            {
                leftArray[i] = array[i];
            }

            for (int i = mid; i < array.Length; i++)
            {
                rightArray[i - mid] = array[i];
            }

            mergeSort(leftArray);
            mergeSort(rightArray);
            Merge(array, leftArray, rightArray);
        }
    }

    private static void Merge(int[] array, int[] leftArray, int[] rightArray)
    {
        int i = 0;
        int j = 0;
        int k = 0;

        while (i < leftArray.Length && j < rightArray.Length)
        {
            if (leftArray[i] < rightArray[j])
            {
                array[k] = leftArray[i];
                i++;
            }
            else
            {
                array[k] = rightArray[j];
                j++;
            }

            k++;
        }

        while (i < leftArray.Length)
        {
            array[k] = leftArray[i];
            i++;
            k++;
        }

        while (j < rightArray.Length)
        {
            array[k] = rightArray[j];
            j++;
            k++;
        }
    }


    public static void Main()
    {
        mergeSort(ARRAY);
        if(debug){
            for(int i = 0; i < ARRAY.Length; i++){
                Console.Write(ARRAY[i] + " ");
            }
        }
        
    }
}